betty/job/threading.py

```diff
@@ -0,0 +1,160 @@
+from __future__ import annotations
+
+from asyncio import run
+from dataclasses import dataclass
+from queue import Queue
+from threading import Semaphore, Event
+from typing import (
+    final,
+    MutableSequence,
+    Callable,
+    Awaitable,
+    Self,
+    ParamSpec,
+    Any,
+    TypeAlias,
+    Concatenate, TypeVar,
+)
+
+from typing_extensions import override
+
+from betty.concurrent import Lock, AsynchronizedLock
+from betty.job import Scheduler
+
+_JobArgT = TypeVar("_JobArgT")
+_JobP = ParamSpec("_JobP")
+_QueueItem: TypeAlias = tuple[
+    Callable[[_JobArgT, ...], Awaitable[None]], tuple[Any], dict[str, Any]
+]
+
+
+@dataclass
+class _SharedState:
+    lock: Lock
+    queues: MutableSequence[Queue[_QueueItem]]
```

Comment:
Batches can be injected in the middle. A sequence won't cut it because the order can change